packages/test/test-end-to-end-tests/src/test/summarizeRefreshLatestSummary.spec.ts

```diff
@@ -53,5 +48,79 @@ describeCompat(
 			await createSummarizer(provider, container, undefined, summaryVersion);
 			await provider.ensureSynchronized();
 		});
+
+		itExpects(
+			"The summarizing client will immediately refresh its own summaries",
+			[
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_start" },
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_end" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_start" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_end" },
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_start" },
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_end" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_start" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_end" },
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_start" },
+				{ eventName: "fluid:telemetry:Summarizer:Running:Summarize_end" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_start" },
+				{ eventName: "fluid:telemetry:SummarizerNode:refreshLatestSummary_end" },
+			],
+			async () => {
+				const summaryConfig: ISummaryConfiguration = {
+					...DefaultSummaryConfiguration,
+					...{
+						maxTime: 5000 * 12,
+						maxAckWaitTime: 120000,
+						maxOps: 1,
+						initialSummarizerDelayMs: 0,
+					},
+				};
+				const testContainerConfig: ITestContainerConfig = {
+					runtimeOptions: {
+						summaryOptions: {
+							summaryConfigOverrides: summaryConfig,
+						},
+					},
+					loaderProps: {
+						configProvider: mockConfigProvider({
+							"Fluid.Summarizer.immediatelyRefreshLatestSummaryAck": true,
+						}),
+					},
+				};
+				const container = await provider.makeTestContainer(testContainerConfig);
+				const container2 = await provider.loadTestContainer(testContainerConfig);
+				const dataObject = (await container.getEntryPoint()) as ITestDataObject;
+				const dataObject2 = (await container2.getEntryPoint()) as ITestDataObject;
+				await waitForContainerConnection(container);
+				await waitForContainerConnection(container2);
+
+				// The first summary needs to be generated otherwise the summarizer will only generate one summary.
+				dataObject._root.set(`first`, "op");
+				await provider.ensureSynchronized();
+
+				const array: string[] = [];
+				container.on("op", (op) => {
+					if (op.type === MessageType.SummaryAck) {
+						array.push((op as ISummaryAckMessage).contents.handle);
+					}
+					if (op.type === MessageType.SummaryNack) {
+						throw new Error("Unexpected summary nack");
+					}
+				});
+
+				// Sending ops from two different clients/containers causes the heuristics summarizer to handle ops twice, which causes the summarizer to summarize before acking its first summary.
```

Comment:
There is no validation that the second summarize is triggered before the first one is acked. You should add that validation somehow otherwise, it makes the test undetereministic. It may not be testing what is claimed here as everything is async.

Reply:
I didn't know exactly how to add in that validation without just starting to cast to the actual internal runtime implementation. What needs to happen is that the running summarizer listens to every op and figures out if it needs to run the summary heuristics runner, so I set the max ops threshold very low (1 op). Then during the summary itself, there need to be unprocessed ops that sit in the inbound queue that again trigger the heuristics runner from the `afterSummaryAction` (or a timer, but ops is simpler). I have two containers send different ops, because when I try to do it all in one container, maybe all the ops get batched into one op (or are processed together) which doesn't trigger the summary. I'm not exactly sure if what's the case, but when I did try it with 1 container I could not effect the same result. The second op from the second container should trigger the 2nd summary as it sits in the queue unprocessed as the run summary is triggered synchronously. I'm not in love with my repro, but if there's a better way to implement it and write it so it's more testable, I'm all ears.